pkg/resources/opnicluster/services.go

```diff
@@ -600,19 +607,192 @@ func (r *Reconciler) gpuCtrlDeployment() (runtime.Object, reconciler.DesiredStat
+					Kind:       r.opniCluster.Kind,
+					Name:       r.opniCluster.Name,
+					UID:        r.opniCluster.UID,
+				},
+			},
+			Labels: labels,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: labels,
+			Ports: []corev1.ServicePort{
+				{
+					Port:       8000,
+					Name:       "metrics",
+					TargetPort: intstr.FromInt(8000),
+				},
+			},
+		},
+	}
+	return service, deploymentState(r.opniCluster.Spec.Services.Metrics.Enabled), nil
+}
+
+func (r *Reconciler) metricsServiceMonitor() (runtime.Object, reconciler.DesiredState, error) {
+	labels := r.serviceLabels(v1beta1.MetricsService)
+	serviceMonitor := &monitoringv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      labels[resources.AppNameLabel],
+			Namespace: r.opniCluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: r.opniCluster.APIVersion,
+					Kind:       r.opniCluster.Kind,
+					Name:       r.opniCluster.Name,
+					UID:        r.opniCluster.UID,
+				},
+			},
+			Labels: labels,
+		},
+		Spec: monitoringv1.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					Port:          "metrics",
+					ScrapeTimeout: "5s",
+				},
+			},
+		},
+	}
+	if r.opniCluster.Spec.Services.Metrics.PrometheusReference == nil {
+		return serviceMonitor, reconciler.StateAbsent, nil
+	}
+	return serviceMonitor, deploymentState(r.opniCluster.Spec.Services.Metrics.Enabled), nil
+}
+
+func (r *Reconciler) metricsPrometheusRule() (runtime.Object, reconciler.DesiredState, error) {
+	labels := r.serviceLabels(v1beta1.MetricsService)
+	prometheusRule := &monitoringv1.PrometheusRule{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", labels[resources.AppNameLabel], r.generateSHAID()),
+			Namespace: r.opniCluster.Namespace,
+			Labels:    labels,
+		},
+		Spec: monitoringv1.PrometheusRuleSpec{
+			Groups: []monitoringv1.RuleGroup{
+				{
+					Name: "opni-rules",
+					Rules: []monitoringv1.Rule{
+						{
+							Alert: "ClusterCpuUsageAnomalous",
+							Expr:  intstr.FromString(`cpu_usage{value_type="is_alert"} >= 1`),
+							Annotations: map[string]string{
+								"summary":     "Anomalous CPU usage in cluster",
+								"description": "Opni has calculated an alert score >= 5 for cluster CPU usage",
+							},
+						},
+						{
+							Alert: "ClusterMemoryUsageAnomalous",
+							Expr:  intstr.FromString(`memory_usage{value_type="is_alert"} >= 1`),
+							Annotations: map[string]string{
+								"summary":     "Anomalous memory usage in cluster",
+								"description": "Opni has calculated an alert score >= 5 for cluster memory usage",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	// Fetch Prometheus resource to calculate namespace and match labels for rules
+	if r.opniCluster.Spec.Services.Metrics.PrometheusReference != nil {
+		prometheus := &monitoringv1.Prometheus{}
+		err := r.client.Get(r.ctx, types.NamespacedName{
```

Comment:
you can use `client.ObjectKeyFromObject()` instead of manually filling out the NamespacedName

Reply:
This doesn't implement client.Object so we can't use that.